Userland/Libraries/LibWeb/CSS/StyleComputer.cpp

```diff
@@ -1234,8 +1235,76 @@ static NonnullRefPtr<StyleValue const> interpolate_box_shadow(DOM::Element& elem
+                return NumericBaseTypeAndDefault { CSSNumericType::BaseType::Time, default_time_value };
+            }
+            case StyleValue::Type::Frequency: {
+                static auto default_frequency_value = FrequencyStyleValue::create(Frequency::make_hertz(0));
+                return NumericBaseTypeAndDefault { CSSNumericType::BaseType::Frequency, default_frequency_value };
+            }
+            case StyleValue::Type::Percentage: {
+                static auto default_percentage_value = PercentageStyleValue::create(Percentage { 0.0 });
+                return NumericBaseTypeAndDefault { CSSNumericType::BaseType::Percent, default_percentage_value };
+            }
+            default:
+                return {};
+            }
+        };
+
+        static constexpr auto to_calculation_node = [](StyleValue const& value) -> NonnullOwnPtr<CalculationNode> {
+            if (value.is_percentage())
+                return NumericCalculationNode::create(value.as_percentage().percentage());
+            if (value.is_length())
+                return NumericCalculationNode::create(value.as_length().length());
+            if (value.is_angle())
+                return NumericCalculationNode::create(value.as_angle().angle());
+            if (value.is_frequency())
+                return NumericCalculationNode::create(value.as_frequency().frequency());
+            if (value.is_time())
+                return NumericCalculationNode::create(value.as_time().time());
+            VERIFY_NOT_REACHED();
+        };
+
+        auto from_base_type_and_default = numeric_base_type_and_default(from);
+        auto to_base_type_and_default = numeric_base_type_and_default(to);
+
+        if (from_base_type_and_default.has_value() && to_base_type_and_default.has_value() && (from_base_type_and_default->base_type == CSSNumericType::BaseType::Percent || to_base_type_and_default->base_type == CSSNumericType::BaseType::Percent)) {
+            // This is an interpolation from a numeric unit to a percentage, or vice versa. The trick here is to
+            // interpolate two separate values. For example, consider an interpolation from 30px to 80%. It's quite
+            // hard to understand how this interpolation work, but if instead we rewrite the values as "30px + 0%" and
+            // "0px + 80%", then it is very simple to understand; we just interpolate each component separately.
+
+            auto interpolated_from = interpolate_value(element, from, from_base_type_and_default->default_value, delta);
+            auto interpolated_to = interpolate_value(element, to_base_type_and_default->default_value, to, delta);
+
+            Vector<NonnullOwnPtr<CalculationNode>> values;
+            values.ensure_capacity(2);
+            values.unchecked_append(to_calculation_node(interpolated_from));
+            values.unchecked_append(to_calculation_node(interpolated_to));
```

Comment:
IIRC you can construct a vector from a set of values, like:

```suggestion
            Vector<NonnullOwnPtr<CalculationNode>> values {
                to_calculation_node(interpolated_from),
                to_calculation_node(interpolated_to),
            };
```

Reply:
You would think that (as I did originally), but unfortunately this doesn't work, for a reason I don't really understand:

```
/home/matt/code/serenity/Meta/Lagom/../../AK/Vector.h:66:13: error: no matching member function for call to 'unchecked_append'
   66 |             unchecked_append(item);
      |             ^~~~~~~~~~~~~~~~
/home/matt/code/serenity/Userland/Libraries/LibWeb/CSS/StyleComputer.cpp:1301:52: note: in instantiation of member function 'AK::Vector<AK::NonnullOwnPtr<Web::CSS::CalculationNode>>::Vector' requested here
 1301 |             Vector<NonnullOwnPtr<CalculationNode>> values { to_calculation_node(interpolated_from), to_calculation_node(interpolated_to) };
      |                                                    ^
/home/matt/code/serenity/Meta/Lagom/../../AK/Vector.h:279:24: note: candidate template ignored: constraints not satisfied [with U = const AK::NonnullOwnPtr<Web::CSS::CalculationNode> &]
  279 |     ALWAYS_INLINE void unchecked_append(U&& value)
      |                        ^
/home/matt/code/serenity/Meta/Lagom/../../AK/Vector.h:280:14: note: because 'CanBePlacedInsideVector<const AK::NonnullOwnPtr<Web::CSS::CalculationNode> &>' evaluated to false
  280 |     requires(CanBePlacedInsideVector<U>)
      |              ^
/home/matt/code/serenity/Meta/Lagom/../../AK/Vector.h:290:24: note: candidate function not viable: requires 2 arguments, but 1 was provided
  290 |     ALWAYS_INLINE void unchecked_append(StorageType const* values, size_t count)
      |                        ^                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```

I think because they're `NonnullOwnPtr`s, they would have to be `move`d out of the `std::initializer_list` in the constructor? But I'm not sure

Reply:
Discussed this on the Discord, looks like initializer_list doesn't work at all with non-copyable types, so fixing this isn't possible.